### PR TITLE
Fix inference when tuple names differ

### PR DESCRIPTION
**Pull request by jcouv**
### Customer scenario
Use a lambda in a method type inference, but have the element names between the lambda and the `Func<... tuple ...>` differ. This will fail the inference.
There is a similar bug with difference in dynamic-ness (ie. `dynamic` vs. `object`), which this PR also fixes.

### Bugs this fixes
Fixes https://github.com/dotnet/roslyn/issues/24781

### Workarounds, if any
Specify type arguments to avoid relying on type inference.

### Performance impact
Low. One line fix.

### Is this a regression from a previous update?
No.

### Root cause analysis
This bug has been there since the tuples feature shipped. It stems from using the `==` comparison between type symbols (which compares everything), instead of the `Equals` method with comparison options.
We have an issue to remove usage of this `==` (and `!=`) in the compiler code: https://github.com/dotnet/roslyn/issues/22067

### How was the bug found?
Reported by customer.

### CODE REVIEW

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs [hunk 1]
```diff
@@ -24253,5 +24253,67 @@ public ValueTuple(T1 item1, T2 item2)
...
+        var case3 = impl.Do(((int x, int y) a) => a.x * a.y);
+    }
+}";
+
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics();
+
+            var tree = comp.SyntaxTrees.First();
+            var model = comp.GetSemanticModel(tree);
+            var varType = tree.GetCompilationUnitRoot().DescendantNodes().OfType<LocalDeclarationStatementSyntax>().Single().Declaration.Type;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs]: >var varType [](start = 12, length = 11)

it is hard to reason what this node is.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs [hunk 2]
```diff
@@ -24253,5 +24253,67 @@ public ValueTuple(T1 item1, T2 item2)
...
+            Assert.Equal("IResult<System.Int32>", model.GetTypeInfo(varType).Type.ToTestDisplayString());
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs]: >GetTypeInfo [](start = 56, length = 11)

It we are testing method type argument inference, why not assert the method symbol?

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs [hunk 3]
```diff
@@ -24253,5 +24253,67 @@ public ValueTuple(T1 item1, T2 item2)
...
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs]: >Assert.Equal("IResult<System.Int32>", model.GetTypeInfo(varType).Type.ToTestDisplayString()); [](start = 12, length = 93)

Same comments

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs [hunk 4]
```diff
@@ -24253,5 +24253,67 @@ public ValueTuple(T1 item1, T2 item2)
...
+    IResult<TReturn> Do<TReturn>(Func<T, TReturn> fn);
+}
+
+public interface IResult<TReturn>
+{
+}
+
+public class Class1
+{
+    public void Test1(IDo<(int, int)> impl)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs]: >(int, int) [](start = 26, length = 10)

Consider also testing scenario when there are names here and they do not match those specified in a lambda

File: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb [hunk 5]
```diff
@@ -21482,6 +21482,40 @@ End Module
...
+</compilation>, additionalRefs:=s_valueTupleRefs)
+            comp.AssertTheseDiagnostics()
+
+            Dim tree = comp.SyntaxTrees(0)
+
+            Dim model = comp.GetSemanticModel(tree)
+            Dim nodes = tree.GetCompilationUnitRoot().DescendantNodes()
+            Dim case3 = nodes.OfType(Of LocalDeclarationStatementSyntax)().Single().Declarators.Single().Names.Single()
+
+            Assert.Equal("case3 As IResult(Of System.Int32)", model.GetDeclaredSymbol(case3).ToTestDisplayString())
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb]: >Assert.Equal("case3 As IResult(Of System.Int32)", model.GetDeclaredSymbol(case3).ToTestDisplayString()) [](start = 12, length = 103)

Same comments

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1)

**Comment by jcouv** [comment]
@jaredpar for ask-mode approval for 15.7. Thanks

The CI errors are not related to this PR (PDBUsingTests issue) and I'll re-run tests once the workaround is merged.

**Comment by jaredpar** [comment]
approved

**Comment by jcouv** [comment]
test windows_debug_unit32_prtest please

**Comment by jcouv** [comment]
test windows_release_unit64_prtest  please

**Comment by jcouv** [comment]
`Microsoft.CodeAnalysis.Editor.UnitTests.IntelliSense.VisualBasicCompletionCommandHandlerTests_Projections.TestInObjectCreationExpression` failed in windows_debug_unit32_prtest  ([details](https://ci.dot.net/job/dotnet_roslyn/job/dev15.7.x/job/windows_debug_unit32_prtest/879/))

@dotnet/roslyn-infrastructure FYI I'll re-run

**Comment by jcouv** [comment]
test windows_debug_unit32_prtest  please

**Pull request merged**